### DISCUSSION TITLE
Is there a way to model stimuli with different durations with the <stimdur> variable?

### DISCUSSION THREAD

**Discussion by jungheejung** [cvnlab/GLMsingle#26] [Q&A] [answered]
I have 4 types of stimuli: A, B, C, D. 
Each stimulus has a duration of 1, 4, 5, 4 seconds. 
Can I model stimuli with different durations in GLMsingle?

**Comment by kendrickkay** [comment]
The short answer is "no, not really".  But the long answer requires some more thought.  Can you tell me a little more about the experimental events that had these different durations? 

The issue here is that within some range, the predicted hemodynamic response (via assumption of convolution with a fixed HRF) to neural events of different durations might actually be very similar in the overall timecourse shape and differ primarily in their amplitude.  If that is true, then everything works well, in the sense that you will pick up on the response differences in the amplitude of the timecourse (i.e. the beta weights that the GLM estimates).  For example, see attached figure: the red line is a predicted HRF to a 1-s stimulus and the blue line is a predicted HRF to a 4-s stimulus (both are normalized in amplitude to peak at 1). The timecourses are actually fairly similar, and it might make sense to just model any such data as having a common timecourse.  (There are deeper issues here, also, like, for a given voxel, how do you plan to interpret responses to different conditions if they indeed differ in their durations -- did you want to assume a fixed HRF (say to 1-s duration) and then assume that this should be convolved with a boxcar of duration 4? That fundamentally changes the interpretation...)

![image](https://user-images.githubusercontent.com/2728338/153010053-3218682f-5ef8-4047-9c3e-4cb4e5d36783.png)

**Comment by jungheejung** [comment]
Hello @kendrickkay Thank you for the quick response. 

## 1. To answer your follow-up question on the experimental events
I have four types of events. 

- A is a cue
- B is a rating phase
- C is a stimulus presentation phase
- D is a rating phase

C is the main event of interest, which consists of 3 levels of high/medium/low stimulus intensity.
The plan is to the model A, B, C, D and then extract beta coefficients for C. 

## 2. Regarding the durations
Your simulations are helpful and interesting in that the HRF is fairly identical except for the time shift. However, I posted my question, primarily because the HRF width does differ as a function of stimulus duration (Figure 2C from  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3318970/ posted below). The duration is particularly important for my experiment, as we deliver noxious heat stimuli.

## 3. Follow up question
What would be your recommendation moving forward for stimuli with different durations? 
*I've been using the LSS method via FSL, but with the preprint, I've been willing to switch gears to the GLMsingle toolbox. The ridge regression, the GLMdenoise, and the flexible model fitting with the HRF library are appealing.*

![image](https://user-images.githubusercontent.com/18406041/153031348-35d3e245-525f-4326-97eb-d7b4617acd9a.png)

**Reply by kendrickkay** [reply]
Yes, regarding the figure, it makes sense.  The width of the predicted timecourse will vary with duration; how much really just depends on the precise numbers for the durations.

My understanding is that you are referring to the 1,4,5,4-s durations that you mentioned for your design.

It sounds like you primarily only care about the 5-s (C) event.  If that is the case, what is your strategy for dealing with the A, B, and D events?  I think that sometimes people are happy to approach the modeling by just using a "generic" single regressor to deal with components that are not of interest (same regressor across all trials).  Whether or not that is sensible really depends on your expectations about BOLD responses evoked in your experiment.

If you are trying to derive a trial-wise estimate for A B and D for every trial, that poses major challenges (due to the single-trial nature of this problem).  Are you trying to do that?

If not, one approach is to construct a single regressor whose goal is to capture all of the evoked responses due to A B and D, and dump that into a single regressor included in the model.  Then, assuming that that is a reasonable modeling approach, you have to deal only with the different C events.  Those could be your main "single trials" for which you are trying to estimate a response.

**Comment by jungheejung** [comment, marked as answer]
@kendrickkay Thank you for the quick response. 

That's a great suggestion — model the events-of-non-interest by dumping into one regressor. I'll definitely use that method. 

Relatedly, since I'm now only modeling event (C) as separate regressors, I'd like to know what the correct way to model event (C) would be in GLMsingle. I'm not sure if I'd be using your [example1 (event-related) ](https://github.com/cvnlab/GLMsingle/blob/main/examples/example1.ipynb) vs [example2 (block design)](https://github.com/cvnlab/GLMsingle/blob/main/examples/example2.ipynb). Here's a short description of my design. 

I have a 2 * 3 event-related design: 

- 2 levels of event (A) * 3 levels of event (C)
- We're interested in whether event (C) is modulated as a function of event (A).
- Within one run, there are 2 repetitions of all 6 combinations (A x C).
- There are 6 runs in total. 

In other words, 6 runs * 2 repetition * 2 event (A) * 3 event (C).
I initially started off with example1 (event-related). However, with the repetition of trials, now I wonder if I should use the example2 block design. What would you suggest?

**Reply by kendrickkay** [reply]
I'm not sure you necessarily need to mimic the examples.

In your experiment, you have 6 combinations that you care about (i.e. A x C).  So, I'll call these 6 distinct experimental conditions.

You mention that you have 2 presentations of each of the 6 conditions in each run and 6 runs, so that's 12 trials in total for each condition.

So, one approach is to specify a stimdur of 5 seconds (corresponding to C).  Assuming that your TR is, say, 1 s, then you would just specify the onset of the C trials in a binary matrix form (like 1 0 0 0 0).  Your design matrix would have 6 columns.  Across the 6 runs, you would have 12 onsets for each condition (but these would be spread across 6 different matrices, one matrix per run).

The tricky thing might be to create the regressor that deals with the events-of-no-interest.  In doing so, you would be incorporating assumptions on what you think the BOLD responses would be like for these events.  If you think that the different levels of A might give rise to different BOLD responses to the A trial component, you might want to specifically model that.    Alternatively, maybe you don't want  to separately model the A B C D components?  Is it possible that you might be fine with a single regressor that captures the entire BOLD response to the each entire trial?  If so, that might greatly simplify this situation.

**Reply by jungheejung** [reply]
@kendrickkay I understand your point. Separate design matrices -- one matrix per run which includes the 12 trial onsets. 6 matrices in total. On it.

> The tricky thing might be to create the regressor that deals with the events-of-no-interest. In doing so, you would be incorporating assumptions on what you think the BOLD responses would be like for these events. If you think that the different levels of A might give rise to different BOLD responses to the A trial component, you might want to specifically model that.

Indeed. But I think we'd be back to square one because event (A) is 1 second duration vs event (C) is 5 seconds. Or am I missing something from your suggestion?

> Alternatively, maybe you don't want to separately model the A B C D components? Is it possible that you might be fine with a single regressor that captures the entire BOLD response to the each entire trial? If so, that might greatly simplify this situation.

So your suggestion is to just model event (C) and have A, B, and D dumped into the error term. Do you see any caveats for not modeling certain events? I am more inclined to your earlier suggestion on the nuissance regressor idea, i.e. model A,B,D as nuissance covariates.

**Reply by kendrickkay** [reply]
Hmm..  I'm not sure we are fully making sense; perhaps we need to have a discussion and a sketch to fully resolve these different ideas.    

One idea is for the user (you) to manually create the regressor intended to model events-of-no-interest.  In doing so, you could account for the different component durations. (see approach 3 below).

And, there are several different approaches that could be used:
(1) model the combination of A+B+C+D so that you have one predictor that is intended to "soak up" all of the BOLD responses that result from the trial
(2) model only the C component and essentially ignore the fact that many other things are happening (i.e. dumping them into the error term)
(3) model only the C component and use some generic events-of-no-interest regressor to deal with the other components, but in a way that isn't too flexible (i.e. just one or two generic regressor that captures all trials).
(4) try to model all of the components separately (which may be impossible/too-unstable/too-hard given the high levels of overlap across components).

**Reply by jungheejung** [reply]
@kendrickkay Ah, this summary is great. Thank you for mapping this out. 
Just to recap, we ruled out (4) due to the complexity of the components. My understanding is that we primarily discussed 2) and (3) in the thread. (1) is new to me, but less desirable for now. 

After the discussions with you, I have a clearer idea of how to move forward. I plan to work on (3) with your suggestions. Thank you again!

**Reply by kendrickkay** [reply]
<img width="825" alt="image" src="https://user-images.githubusercontent.com/2728338/153626402-074dc43a-82be-40d5-8e78-b9a7c9a7ad21.png">
I see.  OK, for path (3), I assume you will have to use the "extraregressors" feature, whereby you can include some predictors that will be fit to the data as part of the model.  Currently, it is set up such that these are specified per-run.  And you don't really get back the GLM weights that are associated with these terms.

**Reply by jungheejung** [reply]
Thank you for the pointers. Will use the opt['extra_regressors']. Good to know that I don't get the GLM weights back. 
Thank you again for the helpful Q&A. I'll mark this thread as answered.
